### ISSUE TITLE
Extend TreeNodeWithChangeSupport with Visitor pattern

### ISSUE THREAD

**Issue by Schmoho** [sbmlteam/jsbml#262]
This is a suggestion: It seems to me that the current implementation does not support the idiomatic way to perform tree walks in Java, because the TreeNode interface that is used does not support the Visitor pattern.

That is, doing a tree walk now requires to explicitly handle type dispatch in the walking class, which is kind of painful to do when also accounting for Plugins and Version differences.

I am very uncertain whether my current understanding of JSBML is correct in that regard and also I am kind of shaky on Java patterns, so please feel free to correct me.

Since the TreeNodeWithChangeSupport interface is the central interface which allows to perform generic tree walks on SBMLDocuments, I would suggest to add at least one interfaces`MutatingTreeNodeVisitor` with signatures `public void visit (Species s)` etc.   TreeNodeWithChangeSupport would then have `void accept(MutatingTreeNodeVisitor v)`.

**Comment by draeger** [comment]
Thank you for this interesting suggestion. In my point of view, the `TreeNode` interface that is on the top of JSBML's type hierarchy was placed in the wrong Java package. There should have been a general algorithm or patterns package but instead this important interface is part of the SWING graphical user interface package. Your idea of making the derived interface `TreeNodeWithChangeSupport` more compatible with recent developments seems sound. However, modifications to a top-level generic interface may have many (potentially also unwanted) effects. In addition, an implementation needs to provided in one of the implementing abstract classes (or a default implementation if suitable). Could you elaborate what subsequent changes to the type hierarchy or implementing classes would be needed?

**Comment by dyrpsf** [comment]
Hi @draeger and @Schmoho ,

I would like to work on this implementation. I agree that the Visitor pattern would provide a much more idiomatic and cleaner way to handle tree walks across the SBML structure.

@draeger , regarding your concern about the type hierarchy and potential unwanted effects: I am happy to map out the specific classes that would be affected. My initial thought is to implement the `accept` method within the central `TreeNodeWithChangeSupport` interface and provide a default implementation in a high-level abstract class (like `AbstractSBase`) to minimize the footprint across subclasses.

Before I move forward with a proposal or PR, I have two questions:
1. Would you prefer separate interfaces for read-only and mutating visitors (e.g., `TreeNodeVisitor` and `MutatingTreeNodeVisitor`), or a single unified visitor interface?
2. Are there specific areas in the SWING-related package hierarchy where you are most concerned about side effects?

I am looking forward to your guidance on this so I can ensure the implementation aligns with the project's long-term design.

**Comment by draeger** [comment]
Thank you for bringing this up again, @dyrpsf.

Let me first clarify my earlier concern about `TreeNode`. My point was not that I expect Swing itself to cause any particular technical problem here. Rather, my concern was the design of the Java class hierarchy: JSBML depends on `javax.swing.tree.TreeNode`, even though it does not use Swing as a GUI framework. When JSBML was originally designed, this was a pragmatic choice because Java did not provide a more suitable standard interface for representing generic tree-like structures. Using `TreeNode` enabled recursive traversal of the entire SBML data structure, compatible with standard Java. This has always been one of the strengths of JSBML compared with libSBML, where such generic recursion over the whole model structure is not available in the same way, at least not in its Java bindings.

In hindsight, however, it would have been cleaner if JSBML had not depended on a Swing interface for such a fundamental model abstraction. That said, I do not think this is a major practical limitation today. If Swing is unavailable in a particular environment, the relevant interface could simply be provided alongside JSBML or replaced with a JSBML-specific equivalent.

Regarding the visitor API, I am not sure that a separate mutating visitor is necessary at first. Since `SBase` is the central abstraction in JSBML, a visitor with a method such as

```
void visit(SBase sbase);
```

would already cover the most important use case. Also, JSBML objects are mutable anyway, so a distinction between a read-only visitor and a mutating visitor would not by itself prevent modifications unless the API exposed immutable views, which JSBML currently does not.

The remaining question is how to handle child nodes that are not instances of SBase, such as annotations or other auxiliary objects. For this reason, it may be useful to add a fallback method based on JSBML’s own tree abstraction, for example:

```
void visit(TreeNodeWithChangeSupport node);
```

This would be more specific than using the most general `Object` and would still allow traversal of nodes in the JSBML tree that are not represented by a more specific SBase implementation. The common case could therefore remain simple via `visit(SBase sbase)`, while `visit(TreeNodeWithChangeSupport node)` would cover the remaining JSBML tree nodes.

It might also be worth considering a generic visitor interface with a return type, for example, `TreeNodeVisitor<T>`, so that visit methods can return values when needed. In cases where no return value is required, `Void` could simply be used as the type parameter. This approach might provide sufficient flexibility while keeping the overall API compact and avoiding the need for multiple separate visitor interfaces.

**Comment by dyrpsf** [comment]
Thank you for the detailed clarification, @draeger ! That makes complete sense. A single, unified generic interface is much cleaner, and using `visit(SBase)` alongside a fallback `visit(TreeNodeWithChangeSupport)` elegantly handles both the core elements and the auxiliary nodes.

Based on your design, the foundation would look something like this:

```java
public interface TreeNodeVisitor<T> {
    
    // Primary traversal for core SBML components
    T visit(SBase sbase);
    
    // Fallback traversal for auxiliary nodes (annotations, etc.)
    T visit(TreeNodeWithChangeSupport node);
}
```
This, combined with your suggestion on PR #305 to use the new `AntimonySerializer` as the primary test-case for this traversal, creates a fantastic development roadmap.

Does this core interface structure look good to you as a starting point? If so, I can begin mapping out the implementation across the abstract classes!

**Comment by draeger** [comment]
This looks fine to me. My suggestion would be to implement that and to see how useful it is in #305 - unless @Schmoho has any particular further recommendation.

In my humble opinion, a practical example like #305 may help us identify necessary adjustments later on (if any).

**Comment by dyrpsf** [comment]
Thank you for the clarification, @draeger! A unified generic interface with a fallback method makes perfect sense.

It seems like @Schmoho might have a busy schedule right now. To keep the momentum going, I will go ahead and start mapping out this implementation across the abstract classes based on this structure. If Schmoho has any further guidance or recommendations later on, I am more than happy to adapt and update the implementation accordingly!

I will open a Draft PR once the initial scaffolding is up so we can evaluate it alongside the AntimonySerializer.

**Comment by dyrpsf** [comment]
Quick update: I have just opened PR #312, which implements the foundational `TreeNodeVisitor` architecture we discussed! 

Since JSBML strictly enforces Java 7 compatibility, we couldn't use Java 8 `default` interface methods. To ensure the contract is perfectly satisfied without breaking any legacy architecture, I explicitly injected the `<T> T accept(TreeNodeVisitor<T> visitor)` method into the 18 required core base classes and math nodes.

Whenever you have a moment, please take a look at the scaffolding in #312. Once this structure is approved and merged, we can start adapting tools like the `AntimonySerializer` to actively use this new visitor pattern!

**Comment by Schmoho** [comment]
I would make this contingent on addressing #313 - this repo needs way more tests, and in particular, this implementation needs way more tests.

**Comment by dyrpsf** [comment]
Just a quick update for anyone tracking this feature: the scaffolding in PR #312 is fully updated and ready for final review! 

The mock visitor tests are in place, and the `accept` method implementations have been successfully rolled out across the foundational abstract classes to ensure all extension plugins compile cleanly.

**Comment by draeger** [comment]
Just a remark on JSBML being bound to Java 7: I think the conservative approach has clear advantages. By sticking to a stable, older Java version, JSBML benefits from Java’s backward compatibility: newer software can still use older Java libraries, while raising the minimum required Java version may prevent some existing downstream projects from upgrading easily.

That being said, Java 8 is a long-term support release, whereas Java 7 is not. Also, Java 7 dates back to 2011, whereas Java 8 was released in 2014. In my humble opinion, after more than a decade, moving at least to Java 8 could be considered reasonable.

The next LTS release would be Java 11 (2018), which might also be old enough by now to assume compatibility with most actively maintained software.

This might be better discussed in a separate issue, though. If we decide to move to a newer Java baseline at some point, it would probably make sense to first release a stable final version that remains fully Java 7 compatible.

**Comment by dyrpsf** [comment]
> Just a remark on JSBML being bound to Java 7: I think the conservative approach has clear advantages. By sticking to a stable, older Java version, JSBML benefits from Java’s backward compatibility: newer software can still use older Java libraries, while raising the minimum required Java version may prevent some existing downstream projects from upgrading easily.
> 
> That being said, Java 8 is a long-term support release, whereas Java 7 is not. Also, Java 7 dates back to 2011, whereas Java 8 was released in 2014. In my humble opinion, after more than a decade, moving at least to Java 8 could be considered reasonable.
> 
> The next LTS release would be Java 11 (2018), which might also be old enough by now to assume compatibility with most actively maintained software.
> 
> This might be better discussed in a separate issue, though. If we decide to move to a newer Java baseline at some point, it would probably make sense to first release a stable final version that remains fully Java 7 compatible.

That makes a lot of sense! Working through the Java 7 constraints for the Visitor pattern definitely highlighted the importance of backward compatibility in a project of this scale. Moving to an LTS release like Java 8 or 11 in the future would definitely streamline things and open up newer language features. 

I can go ahead and open a separate issue to track the discussion on upgrading the Java baseline. That way, we have a dedicated space to plan the transition and map out the final Java 7 stable release!
